### PR TITLE
Implement x/oracle for JUNO

### PR DESCRIPTION
**Pull request by vuong177**
- Implement x/oracle for JUNO

### CODE REVIEW

File: price-feeder/README.md
Comment (adamewozniak) [inline price-feeder/README.md:10]: All of these are differences from the terra fork to the umee fork, everything else here looks the same

Comment (vuong177) [reply price-feeder/README.md:10]: We'll add Juno swap as price provider. Yep it's implemented from Umee.

File: x/oracle/abci.go
Comment (codehans) [inline x/oracle/abci.go:66]: this line needs to be >=, otherwise if you get >67% VP voting for a non-whitelisted asset then everyone gets a miss

**Review by faddat — DISMISSED** [review]

File: price-feeder/README.md
Comment (adamewozniak) [reply price-feeder/README.md:13]: ty for clarifying in docs @vuong177 💯

**Comment by faddat** [comment]
The gov prop for the oracle has passed:

https://www.mintscan.io/juno/proposals/57

We should look into getting this merged for v12.

**Comment by faddat** [comment]
Going through the error logs on the latest commit, I notice that we are having issues subscribing to the binance websocket. 

Kinda :red_flag: to me.

**Review by faddat — DISMISSED** [review]

**Review by faddat — APPROVED** [review]
lgtm

**Review by faddat — DISMISSED** [review]

**Review by faddat — DISMISSED** [review]

**Review by faddat — DISMISSED** [review]

**Review by faddat — DISMISSED** [review]

**Comment by faddat** [comment]
@vuong177 -- I'd like to get this merged, so that further PR's target the main branch.  This branch is getting huge.

**Review by faddat — APPROVED** [review]

**Comment by Reecepbcups** [comment]
I find it hard to stop working... anyways an easier way to review is probably also to compare base branch umee to the changes made here in juno, so I made this repo to do just that. 
Ill review this after the holiday & approve provided no major issues are found.

https://github.com/Reecepbcups/compare-oracle-changes/pull/1/files

**Comment by faddat** [comment]
@Reecepbcups nice review method.

**Review by faddat — APPROVED** [review]

File: x/oracle/types/asset.go
Comment (Reecepbcups) [inline x/oracle/types/asset.go:13]: This is dig channel-37 not atom

```
junod q ibc-transfer denom-trace ibc/6842C591DC4588411A565C9FF650FB15A17DFE3F0A43201E8141E4D14B8D171A
denom_trace:
  base_denom: udig
  path: transfer/channel-37
```

ATOM channel-1 is `ibc/C4CFF46FD6DE35CA4CF4CE031E643C8FDC9BA4B99AE598E9B0ED98FE3A2319F9` for juno.

File: x/oracle/keeper/history_test.go
Comment (Reecepbcups) [inline x/oracle/keeper/history_test.go:577]: Doesn't pass test due to `no values in range`. not sure why this is not coming up on the CI test

**Review by Reecepbcups — CHANGES_REQUESTED** [review]

File: price-feeder/oracle/client/tx.go
Comment (Reecepbcups) [reply price-feeder/oracle/client/tx.go:38]: update - we bypass globalfee now

File: proto/juno/oracle/v1/query.proto
Comment (faddat) [reply proto/juno/oracle/v1/query.proto:34]: "regen files?"

Comment (faddat) [reply proto/juno/oracle/v1/query.proto]: quality finding

File: Makefile
Comment (faddat) [reply Makefile:6]: pretty sure it's simapp for reasons.  Could be wrong though.

@vuong177 ?

File: app/decorators/min_commission.go
Comment (faddat) [reply app/decorators/min_commission.go:13]: Not really sure what you mean but if you'd like to do it in an alternative style, go for it I figure.

This just seems to be declaring a bool to use and... allowing it to be used from decorators.DefaultIsAppSimulation?

File: x/oracle/keeper/history_test.go
Comment (faddat) [reply x/oracle/keeper/history_test.go:577]: the test is an error check?

File: x/oracle/keeper/proposal_handler.go
Comment (faddat) [reply x/oracle/keeper/proposal_handler.go:135]: hi @vuong177 what do you think?

File: proto/juno/oracle/v1/query.proto
Comment (Reecepbcups) [reply proto/juno/oracle/v1/query.proto]: Meant: make proto-all 
So it remakes auto gen file

File: Makefile
Comment (Reecepbcups) [reply Makefile:6]: It’s simapp for a Simulation app, but should be APP for Juno, as this is not cosmos-sdk similation

File: x/oracle/keeper/history_test.go
Comment (Reecepbcups) [reply x/oracle/keeper/history_test.go:577]: Yes, fails with "go test ./..." right now locally for me

File: app/keepers/keepers.go
Comment (faddat) [reply app/keepers/keepers.go:363]: The typo is the misspelling you mentioned?

File: proto/juno/oracle/v1/query.proto
Comment (faddat) [reply proto/juno/oracle/v1/query.proto]: Ahhhh ok not like Regen networks.

Ok cool.  Resolved.

Btw.... Make proto all doesn't work on Mac again.

So this is:

Make proto all (till it fails)
Make proto gen (this works)

File: Makefile
Comment (faddat) [reply Makefile:6]: Hang on lemme poke gaia

Comment (faddat) [reply Makefile:6]: A winner is you

File: app/keepers/keepers.go
Comment (Reecepbcups) [reply app/keepers/keepers.go:363]: yes

File: x/oracle/client/cli/tx.go
Comment (Reecepbcups) [inline x/oracle/client/cli/tx.go:38]: Bit confusing, took me a few tries. 

Should change operator (which seems like to wants your valopper address) and [feeder] to [new_feeder_address]. The docs will specify but not entirely clear here (as someone who has never run an oracle before)

**Review by Reecepbcups — COMMENTED** [review]
Will also add testing via smart contract query panics on query, but instantiate is fine

```bash
junod q wasm contract-state smart $CONTRACT_ADDR '{"exchange_rate":{"denom":"juno"}}'
#where contract address is the oracle_querier.wasm in scripts/

Error: rpc error: code = InvalidArgument desc = Error calling the VM: Error executing Wasm: Wasmer runtime error: RuntimeError: Aborted: panicked at 'called `Result::unwrap()` on an `Err` value: UnsupportedRequest { kind: "'/cosmos.bank.v1beta1.Query/AllBalances' path is not allowed from the contract" }', /home/vuong/.cargo/registry/src/github.com-1ecc6299db9ec823/cosmwasm-std-1.1.5/src/results/system_result.rs:50:9: query wasm contract failed: invalid request
```

File: x/oracle/types/params.go
Comment (faddat) [reply x/oracle/types/params.go:67]: Don't know

File: app/decorators/min_commission.go
Comment (faddat) [reply app/decorators/min_commission.go:13]: Bump?

Comment (Reecepbcups) [reply app/decorators/min_commission.go:13]: resolved in latest commit

File: x/oracle/types/params.go
Comment (vuong177) [reply x/oracle/types/params.go:37]: this is default, we can handler this by upgrade handler.

File: x/oracle/client/cli/tx.go
Comment (Reecepbcups) [reply x/oracle/client/cli/tx.go]: Resolved in 9413c05585d6f9cd84b3d685f599a882c3d3ec11

File: x/oracle/client/cli/query.go
Comment (Reecepbcups) [reply x/oracle/client/cli/query.go:377]: Fixed description in another commit

File: price-feeder/README.md
Comment (JakeHartnell) [inline price-feeder/README.md:4]:
```suggestion
This is Juno's version of [Umee's price feeder](https://github.com/umee-network/umee/tree/main/price-feeder), intended to work with SDK 0.45.11.
```

**Comment by JakeHartnell** [comment]
Do we have any info about this binary `scripts/oracle/oracle_querier.wasm`? Would be nice to have a link in the README about it: https://github.com/CosmosContracts/juno/pull/329/files#diff-f22554e55833f81001f9eab94c6d57144e48d0d3f8496c10c3199a75db5a83bc

Like with a link to the source code.

File: x/oracle/keeper/history.go
Comment (JakeHartnell) [inline x/oracle/keeper/history.go:14]: Remove TODO? I see some tests for this function.
```suggestion
```

**Review by JakeHartnell — APPROVED** [review]
Tremendous amount of work. Made a few minor comments, and agree with @Reecepbcups's asks.

Feel free to tackle those is follow up tickets or issues? Would love to get this in.

Tremendous amount of work! Thank you! Looking forward to battle testing this in testnet.

**Comment by Reecepbcups** [comment]
I will move my above ask into its own issue so we can open another PR for: **oracle upgrade handler & tweaks**

**Review by Reecepbcups — APPROVED** [review]

**Review by JakeHartnell — APPROVED** [review]

**Pull request merged**
